File: src/main/java/com/vc/medium/L863.scala
package com.vc.medium


class TreeNodeL863(var _value: Int) {
  var value: Int = _value
  var left: TreeNodeL863 = null
  var right: TreeNodeL863 = null
}

object L863 {
  def distanceK(root: TreeNodeL863, target: TreeNodeL863, K: Int): List[Int] = {
    import scala.collection.mutable
    val map = new mutable.HashMap[TreeNodeL863, Int]()
    def find(root: TreeNodeL863): Int = {
      if(root == null) return -1
      if(root == target) {
        map.put(root, 0)
        return 0
      }

      val left = find(root.left)
      if(left >= 0) {
        map.put(root, left + 1)
        return left + 1
      }

      val right = find(root.right)
      if(right >= 0) {
        map.put(root, right + 1)
        return right + 1
      }
      -1
    }
    val res = new mutable.ListBuffer[Int]()
    def dfs(root: TreeNodeL863, length: Int): Unit = {
      if(root != null) {
        var len = length
        if(map.contains(root)) len = map(root)
        if(len == K) res += root.value
        dfs(root.left, len + 1)
        dfs(root.right, len + 1)
      }
    }
    find(root)
    dfs(root, 0)
    res.toList
  }
}

File: src/main/java/com/vc/medium/L863.java
package com.vc.medium;

import com.vc.hard.TreeNode;

import java.util.*;

class L863 {

    private HashMap<TreeNode, TreeNode> parent;

    public List<Integer> distanceK(TreeNode root, TreeNode target, int K) {
        this.parent = new HashMap<>();
        helper(root);

        Queue<TreeNode> q = new LinkedList<>();
        q.offer(target);

        HashSet<TreeNode> visited = new HashSet<TreeNode>();
        while(K > 0) {
            int size = q.size();
            while(size > 0) {
                TreeNode e = q.poll();

                if(visited.contains(e)) {
                    size--;
                    continue;
                }

                visited.add(e);

                if(e.left != null && !visited.contains(e.left)) q.offer(e.left);
                if(e.right != null && !visited.contains(e.right)) q.offer(e.right);
                if(parent.containsKey(e) && !visited.contains(parent.get(e))) q.offer(parent.get(e));

                size--;
            }
            K--;
        }

        List<Integer> res = new ArrayList<>();
        while(!q.isEmpty()) res.add(q.poll().val);
        return res;
    }

    private void helper(TreeNode root) {
        if(root == null) return;

        if(root.left != null) {
            //System.out.println(root.left.val+" parent: "+root.val);
            parent.put(root.left, root);
        }
        if(root.right != null) {
            //System.out.println(root.right.val+" parent: "+root.val);
            parent.put(root.right, root);
        }

        helper(root.left);
        helper(root.right);
    }
}
